File: src/main/java/org/wysko/midis2jam2/instrument/percussion/drumset/HiHat.java
package org.wysko.midis2jam2.instrument.percussion.drumset;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.percussion.CymbalAnimator;
import org.wysko.midis2jam2.midi.MidiNoteOnEvent;

import java.util.List;
import java.util.stream.Collectors;

import static org.wysko.midis2jam2.Midis2jam2.rad;

public class HiHat extends SingleStickInstrument {
	
	private final static int WOBBLE_SPEED = 10;
	
	private final static double DAMPENING = 2;
	
	private final static double AMPLITUDE = 0.25;
	
	final List<MidiNoteOnEvent> hitsToStrike;
	
	final Node topCymbal = new Node();
	
	final Node bottomCymbal = new Node();
	
	final Node wholeHat = new Node();
	
	private final CymbalAnimator animator;
	
	private double animTime;
	
	private HiHatStatus status = HiHatStatus.CLOSED;
	
	protected HiHat(Midis2jam2 context, List<MidiNoteOnEvent> hits) {
		super(context, hits);
		hitsToStrike = hits.stream().filter(h -> h.note == 42 || h.note == 46).collect(Collectors.toList());
		Spatial topCymbalModel = context.loadModel("DrumSet_Cymbal.obj", "CymbalSkinSphereMap.bmp",
				Midis2jam2.MatType.REFLECTIVE, 0.7f);
		Spatial bottomCymbalModel = context.loadModel("DrumSet_Cymbal.obj", "CymbalSkinSphereMap.bmp",
				Midis2jam2.MatType.REFLECTIVE, 0.7f);
		bottomCymbalModel.setLocalRotation(new Quaternion().fromAngles(rad(180), 0, 0));
		
		topCymbal.setLocalTranslation(0, 1.2f, 0);
		topCymbal.attachChild(topCymbalModel);
		bottomCymbal.attachChild(bottomCymbalModel);
		wholeHat.attachChild(topCymbal);
		wholeHat.attachChild(bottomCymbal);
		wholeHat.setLocalScale(1.3f);
		
		wholeHat.setLocalTranslation(0, 0, -14);
		highLevelNode.attachChild(wholeHat);
		highLevelNode.setLocalTranslation(-6, 22, -72);
		highLevelNode.setLocalRotation(new Quaternion().fromAngles(0, rad(90), 0));
		highLevelNode.detachChild(stickNode);
		wholeHat.attachChild(stickNode);
		stickNode.setLocalTranslation(0, 1, 13);
		this.animator = new CymbalAnimator(AMPLITUDE, WOBBLE_SPEED, DAMPENING);
	}
	
	@Override
	public void tick(double time, float delta) {
		animator.tick(delta);
		MidiNoteOnEvent recoil = null;
		while (!hits.isEmpty() && context.file.eventInSeconds(hits.get(0)) <= time) {
			recoil = hits.remove(0);
		}
		
		if (recoil != null) {
			animator.strike();
			wholeHat.setLocalTranslation(0, (float) (-0.7 * velocityRecoilDampening(recoil.velocity)), -14);
			if (recoil.note == 46) {
				status = HiHatStatus.OPEN;
				topCymbal.setLocalTranslation(0, 2, 0);
			} else {
				status = HiHatStatus.CLOSED;
				topCymbal.setLocalTranslation(0, 1.2f, 0);
			}
			animTime = 0;
		}
		topCymbal.setLocalRotation(new Quaternion().fromAngles(status == HiHatStatus.CLOSED ? 0 : animator.rotationAmount(), 0, 0));
		if (animTime != -1) animTime += delta;
		handleStick(time, delta, hitsToStrike);
		
		wholeHat.move(0, 0.025f, 0);
		if (wholeHat.getLocalTranslation().y > 0) {
			Vector3f localTranslation = new Vector3f(wholeHat.getLocalTranslation());
			localTranslation.y = Math.min(localTranslation.y, 0);
			wholeHat.setLocalTranslation(localTranslation);
		}
	}
	
	private enum HiHatStatus {
		CLOSED,
		OPEN
	}
}

File: src/main/kotlin/org/wysko/midis2jam2/instrument/family/percussion/drumset/kit/HiHat.kt
/*
 * Copyright (C) 2024 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.percussion.drumset.kit

import com.jme3.math.FastMath
import com.jme3.math.Vector3f
import com.jme3.scene.Node
import com.jme3.scene.Spatial
import org.wysko.kmidi.midi.event.NoteEvent
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.algorithmic.EventCollector
import org.wysko.midis2jam2.instrument.algorithmic.StickType
import org.wysko.midis2jam2.instrument.algorithmic.Striker
import org.wysko.midis2jam2.instrument.family.percussion.CymbalAnimator
import org.wysko.midis2jam2.instrument.family.percussion.drumset.DrumSetInstrument
import org.wysko.midis2jam2.util.max
import org.wysko.midis2jam2.world.modelR
import kotlin.time.Duration

private val CLOSED_POSITION = Vector3f(0f, 1.2f, 0f)
private val OPEN_POSITION = Vector3f(0f, 2f, 0f)
private const val AMPLITUDE = 0.25
private const val DAMPENING = 2.0
private const val WOBBLE_SPEED = 10.0

/** The hi-hat. */
class HiHat(
    context: Midis2jam2,
    hits: MutableList<NoteEvent.NoteOn>,
    private val noteMapping: HiHatNoteMapping = HiHatNoteMapping.Standard,
) : DrumSetInstrument(context, hits) {
    private val cymbalsNode =
        Node().apply {
            recoilNode.attachChild(this)
            scale(1.3f)
            move(0f, 0f, -13f)
        }

    private val topCymbal: Spatial =
        context.modelR(
            model = "DrumSet_Cymbal.obj",
            texture = "CymbalSkinSphereMap.bmp"
        ).apply {
            localTranslation = CLOSED_POSITION // Start in closed position
            cymbalsNode.attachChild(this)
        }

    init {
        // Add bottom cymbal
        context.modelR(
            model = "DrumSet_Cymbal.obj",
            texture = "CymbalSkinSphereMap.bmp"
        ).apply {
            rotate(FastMath.PI, 0f, 0f) // Rotate upside down
            cymbalsNode.attachChild(this)
        }
    }

    private val stick: Striker =
        Striker(
            context = context,
            strikeEvents = hits.filter { it.note != noteMapping.pedal },
            stickModel = StickType.DRUM_SET_STICK,
        ).apply {
            setParent(recoilNode)
            node.move(0f, 1f, 2f)
        }

    private val cymbalAnimator = CymbalAnimator(topCymbal, AMPLITUDE, WOBBLE_SPEED, DAMPENING)

    private val pedalEventCollector =
        EventCollector(
            context = context,
            events = hits.toList().filter { it.note == noteMapping.pedal },
        )

    init {
        geometry.move(-6f, 22f, -72f)
        geometry.rotate(0f, 1.57f, 0f)
    }

    override fun tick(
        time: Duration,
        delta: Duration,
    ) {
        super.tick(time, delta)

        val stickResults = stick.tick(time, delta)
        stickResults.strike?.let {
            cymbalAnimator.strike()
            topCymbal.localTranslation =
                when (it.note) {
                    noteMapping.open -> OPEN_POSITION
                    else -> {
                        cymbalAnimator.cancel()
                        CLOSED_POSITION
                    }
                }
        }

        val pedalResults = pedalEventCollector.advanceCollectOne(time)
        pedalResults?.let {
            cymbalAnimator.cancel()
            topCymbal.localTranslation = CLOSED_POSITION
        }

        recoilDrum(
            drum = recoilNode,
            velocity = max(stickResults.strike?.velocity ?: 0, pedalResults?.velocity ?: 0),
            delta = delta,
            recoilDistance = -0.7f,
            recoilSpeed = 12f,
        )

        cymbalAnimator.tick(delta)
    }
}

/**
 * The hi-hat note mapping.
 *
 * @property closed The MIDI note for the closed hi-hat.
 * @property open The MIDI note for the open hi-hat.
 * @property pedal The MIDI note for the hi-hat pedal.
 */
sealed class HiHatNoteMapping(
    val closed: Byte,
    val open: Byte,
    val pedal: Byte,
) {
    /** The mapping as seen in most all kits. */
    data object Standard : HiHatNoteMapping(
        closed = 42,
        open = 46,
        pedal = 44,
    )

    /** The mapping as seen in the orchestra kit. */
    data object Orchestra : HiHatNoteMapping(
        closed = 27,
        open = 29,
        pedal = 28,
    )
}
